vortex-array/src/arrays/fixed_size_list/mod.rs

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use std::sync::Arc;
+
+use vortex_dtype::DType;
+use vortex_error::{VortexExpect, VortexResult, vortex_ensure};
+
+use crate::stats::ArrayStats;
+use crate::validity::Validity;
+use crate::{Array, ArrayRef};
+
+mod vtable;
+pub use vtable::{FixedSizeListEncoding, FixedSizeListVTable};
+
+#[cfg(test)]
+mod tests;
+
+/// The canonical encoding for fixed-size list arrays.
+#[derive(Clone, Debug)]
+pub struct FixedSizeListArray {
+    /// The [`DType`] of the fixed-size list.
+    ///
+    /// This type **must** be the variant [`DType::FixedSizeList`].
+    dtype: DType,
+
+    /// The `elements` data array, where each fixed-size list scalar is a _slice_ of the `elements`
+    /// array, and each inner list element is a _scalar_ of the `elements` array.
+    ///
+    /// The fixed-size list scalars (or the elements of the array) are contiguous (regardless of
+    /// nullability for easy lookups), each with equal size in memory.
+    elements: ArrayRef,
+
+    /// The size of each fixed-size list scalar in the array.
+    ///
+    /// We store the size of each fixed-size list in the array as a field for convenience.
+    list_size: u32,
+
+    /// The validity / null map of the array.
+    ///
+    /// Note that this null map refers to the fixed-size list scalars, **not** the elements of the
+    /// _individual_ fixed-size list scalars. The `elements` array will track individual value
+    /// nullability.
+    validity: Validity,
+
+    /// The length of the array.
+    ///
+    /// Note that this is different from the size of each fixed-size list scalar (`list_size`).
+    ///
+    /// The main reason we need to store this (rather than calculate it on the fly via `list_size`
+    /// and `elements.len()`) is because in the degenerate case where `list_size == 0`, we cannot
+    /// use `0 / 0` to determine the length.
+    len: usize,
+
+    /// The stats for this array.
+    stats_set: ArrayStats,
+}
+
+impl FixedSizeListArray {
+    /// Creates a new `FixedSizeListArray`. This is simply a wrapper around [`try_new()`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if the inputs are invalid. See
+    ///
+    /// [`try_new()`]: Self::try_new
+    pub fn new(elements: ArrayRef, list_size: u32, validity: Validity, len: usize) -> Self {
```

Comment:
We should also add a new_unchecked constructor that you use for the implementations of the operations. We maintain the invariants in our implementation and don't need to check them again on construction